common/dynamicconfig/cachedvalue.go

```diff
@@ -0,0 +1,59 @@
+package dynamicconfig
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"go.temporal.io/server/common/log/tag"
+)
+
+// GlobalCachedTypedValue holds a cached value of a GlobalTypedSetting after applying a supplied convert function.
+// This type should be deleted after caching is added to GlobalTypedSettingWithConverter.
+type GlobalCachedTypedValue[T any] struct {
+	ptr    *atomic.Pointer[T]
+	cancel func()
+}
+
+// NewGlobalCachedTypedValue returns a new [GlobalCachedTypedValue] for the given setting, applying the given convert
+// function on change.
+// If convert returns an error the old value will be preserved. If convert never succeeds, Get() will return the zero
+// value of T.
+func NewGlobalCachedTypedValue[T, I any](c *Collection, setting GlobalTypedSetting[I], convert func(I) (T, error)) *GlobalCachedTypedValue[T] {
+	ptr := &atomic.Pointer[T]{}
```

Comment:
e.g. instead of this, do
```suggestion
	cachedValue := new(GlobalCachedTypedValue[T])
```
and then do `cachedValue.ptr.Store(...)` and `cachedValue.cancel = ...` below